### ISSUE TITLE
ndt_mapping execution problem

### ISSUE THREAD

**Issue by dyheggi** [daohu527/ndt_mapping#12]
Hello,Mr.daohu!

I encountered 'segmentation fault' error while running NDT algorithm.I inserted the detection parameter ‘times’

Then I find the map in the lidarprocess function after the second loop_ ptr statements are no longer executed.That is, line 203 in the figure.

I want to know what caused this problem,  I would appreciate it if you could help me!!!

![2022-04-21 18-38-04 的屏幕截图](https://user-images.githubusercontent.com/103110065/164441330-18ab0793-8be0-46dc-9aa2-7e16554b7681.png)

![2022-04-21 18-39-54 的屏幕截图](https://user-images.githubusercontent.com/103110065/164441354-6ce59fd6-1af1-4941-a0fe-3ea497b2d7b3.png)

**Comment by daohu527** [comment]
It looks like this has something to do with the PCL library, I found some introductions online, you can try to remove the nan points first
```c++
  PointCloudPtr scan_ptr(new PointCloud());
  vector<int> index;
  pcl::removeNaNFromPointCloud(*cloud_ptr,*cloud_ptr,index);   // remove nan points
  RangeFilter(cloud_ptr, scan_ptr);
```

**Comment by dyheggi** [comment]
After trying, there was an error in the compilation（./apollo.sh build localization）, which showed that the vector was not declared, and there was a problem with int 

![2022-04-21 22-31-52 的屏幕截图](https://user-images.githubusercontent.com/103110065/164481215-39dc8d58-86c4-4fd0-9600-87d6670e14cf.png)

**Comment by daohu527** [comment]
try
```c++
  std::vector<int> index;
  pcl::removeNaNFromPointCloud(*cloud_ptr,*cloud_ptr,index);
```

**Comment by dyheggi** [comment]
This time, it can be compiled, but there are still the same errors as before

![2022-04-22 09-21-54 的屏幕截图](https://user-images.githubusercontent.com/103110065/164578121-7b8b0b7b-5bf5-49ff-8bc7-c0a4b32bbdbb.png)

**Comment by daohu527** [comment]
If you can, you can send 3-5 frames of data to daohu527@gmail.com, I will have a try

**Comment by dyheggi** [comment]
It has been sent to you（dyheggi@163.com）

**Comment by dyheggi** [comment]
I made the following changes: 

1.In order to prevent the error of intensity, the intensity data type of the parsed PCD file is changed to 'float'

2.The parsed files include not only PCD files, but also fusion, lidar, GNSS and other LOC Txt file, so I don't know if I can run the program normally by sending you only a few PCD files

**Comment by dyheggi** [comment]
I've tried, but I still need those txt files, which can't be executed only with PCD, so if you can, I want to send you the data packet I recorded, or the whole PCD I parsed

**Comment by daohu527** [comment]
@eggiiiiiii I tested it, **apollo5.0 can work normally**, but 6.0 not.

At the moment I don't know what the problem is, it might take some time. You can use apollo5.0 and [r5.0.0](https://github.com/daohu527/ndt_mapping/tree/r5.0.0) to build a map, it doesn't need to change anything, just switch branches and follow the same steps

**Comment by dyheggi** [comment]
Ok, my current computer is apollo5.0, pcl version is 1.10.0, apollo car in the laboratory is apollo5.5.0, and pcl version is 1.8.

I'm going to test whether apollo5.5 version 5.5 can run.

**Comment by dyheggi** [comment]
I installed apollo5.0, move NDT related files to local_ tool the directory, GPU cannot be compiled when entering the docker environment, like this

![2022-04-22 17-41-10 的屏幕截图](https://user-images.githubusercontent.com/103110065/164700188-10d78fd5-d4ff-4e02-838a-6327a66896af.png)

 This situation is like our laboratory car, which is equipped with Apollo Version 5.5

**Comment by daohu527** [comment]
You should check out the `ndt_mapping` in 5.0.0 too.
```
cd modules/localization/msf/local_tool/ndt_mapping
git checkout r5.0.0
```

**Comment by dyheggi** [comment]
Encountered such a problem 

![2022-04-22 21-48-06 的屏幕截图](https://user-images.githubusercontent.com/103110065/164727343-115996b1-0122-4db9-b98b-62e02117b014.png)

**Comment by daohu527** [comment]
you can clone the r5.0.0 branch
```
git clone -b r5.0.0 git@github.com:daohu527/ndt_mapping.git
```

**Comment by dyheggi** [comment]
Sorry, I still can't. the execution code shows that I don't have access rights. 

Downloading the compressed package directly from GitHub and decompressing it will also report an error as before 

![2022-04-23 13-51-11 的屏幕截图](https://user-images.githubusercontent.com/103110065/164880135-bba4fd22-366e-4d1b-8040-314b3024190a.png)

**Comment by dyheggi** [comment]
In addition, let me ask you, what is your PCL version? Mine is 1.10.0. Will this affect the program ?

**Comment by daohu527** [comment]
If you run apollo 5.0 docker, the pcl version is 1.7, If apollo 6.0 the pcl version is 1.10

**Comment by dyheggi** [comment]
I checked my PCL and found that it is still version 1.10. Do I need to install a 1.7 PCL library？

And I still can't copy your NDT_ Mapping code...

**Comment by daohu527** [comment]
I'm confused how you are running apollo, if you are running in docker why do you have this annoyance.

I create a release version, you can try to download [r5.0.0](https://github.com/daohu527/ndt_mapping/releases/tag/r5.0.0)

**Comment by dyheggi** [comment]
Now GPU compilation and NDT compilation are successful.

 However, I have the following problems when decompressing the data package 

![2022-04-26 10-54-23 的屏幕截图](https://user-images.githubusercontent.com/103110065/165210867-e5570cfd-d5d1-42e8-b326-d87502b84eed.png)


Since our car is a 16 wire radar, we will cyber before_ record_ The following changes have been made to the parser, and the "R32" in the command has been changed to "lidar16". The program can run, but it can't run now. Why

![2022-04-26 10-56-48 的屏幕截图](https://user-images.githubusercontent.com/103110065/165210941-d9099149-eba3-4f15-9e4c-937f037aa3d0.png)

**Comment by dyheggi** [comment]
This is our modified file in apollo6 You can run programs on 0, but not on 5.0 

![2022-04-26 11-06-50 的屏幕截图](https://user-images.githubusercontent.com/103110065/165212055-e5a866a8-3f6a-414c-aff0-4500a59d74e2.png)

**Comment by dyheggi** [comment]
Do you know the format of adding traffic light files to high-precision maps? Now I only put the stop line on it, but the traffic lights can't be put up all the time because of the format problem. If you have an edited file, can you send me a copy? I want to refer to the format.

**Comment by daohu527** [comment]
You can ref to [borregas_ave.txt](https://github.com/daohu527/imap/blob/main/imap/data/borregas_ave.txt)'s signal, Note to add `overlap`, signal and lane connect with `overlap`

By the way, this is clearly beyond the scope of this project's discussion. We'd better focus on the function of the project, if you have other questions, you can go to the apollo community for help

**Comment by dyheggi** [comment]
After continuous debugging, I can now run the NDT algorithm and output the map!

In addition, I would like to ask you what is your tool for visualizing PCD?

My PCL tool on Apollo version 6.0 always fails to install in version 5.0...

**Comment by daohu527** [comment]
I used `pcl_viewer` or `cloudcompare`

**Comment by mbing123** [comment]
Hello,Mr.daohu!
my apollo car computer is apollo5.5. And clone the ndt_mapping_r5.0.0.
Now GPU compilation and NDT compilation are successful.
However, I have the following problems when create a map
I want to know what caused this problem, I would appreciate it if you could help me!!!
![Screenshot from 2022-04-30 18-40-11](https://user-images.githubusercontent.com/104626637/166102765-6b5fad60-e986-4c04-942a-9091d24d6911.png)

**Comment by daohu527** [comment]
If there are no other question, I will close this issue and I will look at apollo 6.0 PCL library issue soon

**Comment by daohu527** [comment]
@eggiiiiiii I fixed this issue for now, Because 6.0 can't download go.sdk now, I can't compile it, you can checkout the r6.0.0 and have a try.
```
cd apollo/modules/localization/msf/local_tool/ndt_mapping
git checkout r6.0.0
```

**Comment by dyheggi** [comment]
I'll try! 

Thank you very much for your help!!!
